src/ILCompiler.Compiler/src/IL/Stubs/StartupCode/StartupCodeMainMethod.cs

```diff
@@ -89,9 +89,20 @@ public override MethodIL EmitIL()
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(startup.GetKnownMethod("GetMainMethodArguments", null)));
             }
             codeStream.Emit(ILOpcode.call, emitter.NewToken(_mainMethod));
+
+            // If the main method returns void, return the latched exit code set by Environment.ExitCode.
             if (_mainMethod.Signature.ReturnType.IsVoid)
             {
-                codeStream.EmitLdc(0);
+                MetadataType environmentAugments = Context.SystemModule.GetType("Internal.Runtime.Augments", "EnvironmentAugments");
```

Comment:
I wanted to keep it in a place that has more potential to be shared with Project N. But I can move it if you insist.